services/device-registry-base/src/main/java/org/eclipse/hono/deviceregistry/service/deviceconnection/MapBasedDeviceConnectionService.java

```diff
@@ -43,15 +47,15 @@
  */
 @Repository
 @Qualifier("backend")
-public final class MapBasedDeviceConnectionService implements DeviceConnectionService {
+public class MapBasedDeviceConnectionService implements DeviceConnectionService {
 
     private static final Logger log = LoggerFactory.getLogger(MapBasedDeviceConnectionService.class);
 
     // <tenantId, <deviceId, lastKnownGatewayJson>>
     private final Map<String, Map<String, JsonObject>> lastKnownGatewaysMap = new HashMap<>();
 
     // <tenantId, <deviceId, adapterInstanceIdJson>>
-    private final Map<String, Map<String, JsonObject>> commandHandlingAdapterInstancesMap = new HashMap<>();
+    private final Map<String, Map<String, ExpiringValue<JsonObject>>> commandHandlingAdapterInstancesMap = new HashMap<>();
```

Comment:
it might be easier to use a Caffeine *Cache* instead which already supports limiting size and defining a time-based eviction strategy ...

Reply:
I've committed a change so that a Caffeine Cache is now used, letting it do the eviction.

I've haven't used the size limited feature though (`maximumSize` on cache builder) because that would change the behaviour. When the max size is (about to get) reached on a Caffeine Cache, some existing entry gets evicted. What we want here is to keep the existing entries and prevent addition of new entries if the max size has been reached.